text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+By unifying all these types, we can simplify all of our IO components drastically. Ensuring that all have a unique hash that includes a type-discriminating tag ensures that we can use the same namespace for identifiers.
+
+As a result, our IO layer will be greatly simplified, as well as its usage. Possible gains are easier snapshotting, DB handling and shipping data around in general. Speed-ups through external caching become very real (see [Future possibilities](#future-possibilities)).
+
+Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+(It is recommended to read the [prior art](#prior-art) section beforehand)
+
+Every *object* that our node handles has common properties, namely:
+
+* It can be serialized to binary storage.
+* It has a unique identifying hash that depends solely on its contents.
+* It may reference other objects.
+
+This is common to everything we handle, for example a block will reference a number of deploys, which in turn reference WASM blobs.
+
+### Serialization
+
+Every object can be serialized to be sent across the network, which also means it can be stored in storage. In general, any serialized object is prefixed with a short tag indicating its type, regardless of the chosen serialization format.
+
+This helps disamgiuate between different objects whose serialization may be the same under a given format, and can also be used to avoid conflicts with versioning, should an existing storage be run with a different node version.
+
+The tag should be kept short, one byte should be plenty to cover all of our usecases (we will reserve one value for future expansion if necessary).
+
+### Hashing
+
+Object hashes are always SHA512 hashes, as these are already in use by the execution engine. It is imperative that we standardize on one kind of hash for all objects.
+
+The hash value is derived solely from the serialized representation of the object, including its tag. This has the advantage that hashes can be checked without deserializing the object, even progressively while downloading it from a peer if necessary.
+
+It also allows storage and hash verification by other tools that are unaware of any of the business logic.
+
+### Object references
+
+Any properly typed object in memory provides a method for iterating over all of its *dependencies*. In graph terms this is an iterator over all outgoing edges.
+
+With this functionality, the logic of fetching dependent objects is moved out of the business logic components and into the fetcher. Even "validation" (verifying all references parts are available) happens at this point, this consumes the entire current logic of the block validator.
```

Comment:
```suggestion
With this functionality, the logic of fetching dependent objects is moved out of the business logic components and into the fetcher. Verifying all referenced dependencies are available is subsumed into the fetcher as well. 

_NOTE: this would render the current block validator component obsolete._
```